generators/java/mixin.cjs

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright 2013-2021 the original author or authors from the JHipster project.
+ *
+ * This file is part of the JHipster project, see https://www.jhipster.tech/
+ * for more information.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const { defaults, merge } = require('lodash');
+
+const { GENERATOR_MAVEN, GENERATOR_GRADLE } = require('../generator-list');
+const { requiredConfig, defaultConfig } = require('./config.cjs');
+const { options } = require('./options.cjs');
+const {
+  JAVA_VERSION,
+  JAVA_APP_VERSION,
+  JAVA_COMPATIBLE_VERSIONS,
+  JHIPSTER_BOM_VERSION,
+  BUILD_TOOL,
+  BUILD_TOOL_DEFAULT,
+  BUILD_TOOL_MAVEN,
+  BUILD_TOOL_GRADLE,
+  BUILD_TOOL_PROMPT_CHOICES,
+  PACKAGE_NAME,
+  PRETTIER_JAVA_INDENT,
+} = require('./constants.cjs');
+
+module.exports.mixin = parent =>
+  class extends parent {
+    /**
+     * Load java options constants.
+     */
+    loadJavaOptionsConstants(into = this) {
+      into.BUILD_TOOL_DEFAULT = BUILD_TOOL_DEFAULT;
+      into.BUILD_TOOL_PROMPT_CHOICES = BUILD_TOOL_PROMPT_CHOICES;
+    }
+
+    /**
+     * Register and parse java options.
+     */
+    registerJavaOptions(customOptions) {
+      this.jhipsterOptions(merge({}, options, customOptions));
+    }
+
+    /**
+     * Load required java configs into config.
+     */
+    configureJava() {
+      this.config.defaults(requiredConfig);
+    }
+
+    /**
+     * Load java configs into into.
+     * all variables should be set to into,
+     * all variables should be referred from config,
+     * @param {any} config - config to load config from
+     * @param {any} into - destination context to use default is context
+     */
+    loadJavaConfig(config = this.jhipsterConfig, into = this) {
+      config = defaults({}, config, defaultConfig);
+      into[PACKAGE_NAME] = config[PACKAGE_NAME];
+      into[BUILD_TOOL] = config[BUILD_TOOL];
+      into[PRETTIER_JAVA_INDENT] = config[PRETTIER_JAVA_INDENT];
+    }
+
+    /**
+     * Load derived java configs into fromInto.
+     * @param {any} fromInto - source/destination context
+     */
+    loadDerivedJavaConfig(fromInto = this) {
+      fromInto.javaMainClass = fromInto.getMainClassName(fromInto.baseName);
+      fromInto.packageFolder = fromInto[PACKAGE_NAME].replace(/\./g, '/');
+
+      fromInto.buildToolMaven = fromInto.buildTool === BUILD_TOOL_MAVEN;
```

Comment:
Same logic here: it's counterIntuitive to have build tool properties in the previous chain: at least having a file in the 'buildTool' folder that reference all the methods that are called from here